test/e2e/storage/csimock/csi_honor_pv_reclaim_policy.go

```diff
@@ -0,0 +1,192 @@
+			framework.ExpectNoError(err, "failed to delete PVC %s", pvc.Name)
+
+			ginkgo.By(fmt.Sprintf("Waiting for PV %s to be deleted", pv.Name))
+			err = e2epv.WaitForPersistentVolumeDeleted(ctx, f.ClientSet, pv.Name, framework.Poll, 2*time.Minute)
+			framework.ExpectNoError(err, "failed to wait for PV to be deleted")
+
+			ginkgo.By(fmt.Sprintf("Verifying that the driver received DeleteVolume call for PV %s", pv.Name))
+			gomega.Expect(m.driver.GetCalls(ctx)).To(gomega.ContainElement(gomega.HaveField("Method", gomega.Equal("DeleteVolume"))))
+		})
+
+		ginkgo.It("Static provisioning should honor pv retain reclaim policy", func(ctx context.Context) {
+			m.init(ctx, testParameters{
+				registerDriver:             true,
+				enableHonorPVReclaimPolicy: true,
+				reclaimPolicy:              ptr.To(v1.PersistentVolumeReclaimRetain),
+			})
+			ginkgo.DeferCleanup(m.cleanup)
+
+			sc, pv, pvc := m.createPVPVC(ctx)
+			gomega.Expect(pv.Spec.PersistentVolumeReclaimPolicy).To(gomega.Equal(v1.PersistentVolumeReclaimRetain),
+				"expected PV %s to have reclaim policy %s, got %s", pv.Name, v1.PersistentVolumeReclaimRetain, pv.Spec.PersistentVolumeReclaimPolicy)
+			gomega.Expect(pv.Annotations).NotTo(gomega.HaveKeyWithValue(storagehelpers.AnnDynamicallyProvisioned, sc.Provisioner), "expected PV %s to not have annotation %s", pv.Name, storagehelpers.AnnDynamicallyProvisioned)
+
+			ginkgo.By(fmt.Sprintf("Verifying that the PV %s does not have finalizer %s after creation", pv.Name, storagehelpers.PVDeletionProtectionFinalizer))
+			gomega.Consistently(ctx, framework.GetObject(f.ClientSet.CoreV1().PersistentVolumes().Get, pv.Name, metav1.GetOptions{})).
+				WithPolling(framework.Poll).WithTimeout(framework.ClaimProvisionTimeout).ShouldNot(gomega.HaveField("Finalizers",
+				gomega.ContainElement(storagehelpers.PVDeletionProtectionFinalizer)), "failed to wait for PV to not have finalizer %s", storagehelpers.PVDeletionProtectionFinalizer)
+
+			ginkgo.By(fmt.Sprintf("Deleting PV %s", pv.Name))
+			err := f.ClientSet.CoreV1().PersistentVolumes().Delete(ctx, pv.Name, metav1.DeleteOptions{})
+			framework.ExpectNoError(err, "failed to delete PV %s", pv.Name)
+
+			ginkgo.By(fmt.Sprintf("Deleting PVC %s", pvc.Name))
+			err = f.ClientSet.CoreV1().PersistentVolumeClaims(pvc.Namespace).Delete(ctx, pvc.Name, metav1.DeleteOptions{})
+			framework.ExpectNoError(err, "failed to delete PVC %s", pvc.Name)
+
+			ginkgo.By(fmt.Sprintf("Waiting for PV %s to be deleted", pv.Name))
+			err = e2epv.WaitForPersistentVolumeDeleted(ctx, f.ClientSet, pv.Name, framework.Poll, 2*time.Minute)
+			framework.ExpectNoError(err, "failed to wait for PV to be deleted")
+
+			ginkgo.By(fmt.Sprintf("Verifying that the driver did not receive DeleteVolume call for PV %s", pv.Name))
+			gomega.Expect(m.driver.GetCalls(ctx)).NotTo(gomega.ContainElement(gomega.HaveField("Method", gomega.Equal("DeleteVolume"))))
```

Comment:
same as before

Reply:
replied in https://github.com/kubernetes/kubernetes/pull/123151#discussion_r1521257565